feat: verify expected `PeerId` as part of security handshake

denis2glez opened the pull request:
## Description

During a handshake, certificates should be verified in order to allow the connection attempt to be aborted during the security handshake (by security protocols such as TLS or Noise). This requires being able to verify the peer ID after the handshake. Otherwise, we will have abnormal behaviors, such as aborting after completing the handshake and leaving the peer not knowing what went wrong.

Fixes: #2946.
Related: #4307.
Related: #4726.
Related: #882.

## Tasks

- [x] Introduce the `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` traits that should be implemented by transports such as Noise or TLS.
- [x] Create the `Builder::authenticate2` function to accept an upgrade that implements this traits.
- [x] Implement a state machine in`upgrade::secure` that calls the `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` traits instead of `InboundConnectionUpgrade` / `OutboundConnectionUpgrade`.
- [x] Implement the `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` traits for Noise and TLS transports, repectively.
- [ ] Complete the missing parts.
- [ ] Include unit tests.
- [ ] Make appropriate changes to the docs.

## Notes & open questions

denis2glez committed:
feat: MVP verification of the expected `PeerId` as part of a protocol handshake.

denis2glez committed:
chore: fix a typo in the doc comments

thomaseizinger commented on core/src/upgrade.rs:
```diff
@@ -152,3 +154,22 @@ pub trait OutboundConnectionUpgrade<T>: UpgradeInfo {
+    /// Possible error during the handshake.
+    type Error;
+    /// Future that performs the handshake with the remote.
+    type Future: Future<Output = Result<Self::Output, Self::Error>>;
```
```suggestion
    type Future: Future<Output = Result<(PeerId, Self::Output), Self::Error>>;
```

With a specialised upgrade, we can fix the `PeerId` here.

denis2glez replied:
That makes sense, thanks.

thomaseizinger commented on core/src/upgrade/secure.rs:
```diff
@@ -0,0 +1,88 @@
+                    Protocol::P2p(peer_id) => Some(peer_id),
+                    _ => None,
+                })
+                .expect("It should have /p2p as part of the address");
```
Can't panic here, if it is not there we should set it to `None`.

denis2glez replied:
Yes, I forgot about the panic.

thomaseizinger commented on core/src/upgrade/secure.rs:
```diff
@@ -0,0 +1,88 @@
+            Err(UpgradeError::Apply(e))
+        }
+    }
+}
```
I'd honestly just inline these functions I think.

denis2glez replied:
I suppose the result will not be very long.

thomaseizinger commented on core/src/transport/upgrade.rs:
```diff
@@ -99,16 +99,15 @@ where
         Authenticated(Builder::new(
             self.inner.and_then(move |conn, endpoint| Authenticate {
-                inner: upgrade::apply(conn, upgrade, endpoint, version),
+                inner: upgrade::secure(conn, upgrade, endpoint, version).boxed_local(),
```
Why `boxed_local`?

denis2glez replied:
Because the function parameters are not `Send`.

thomaseizinger replied:
I think effectively, they will be so we can just add the necessary constraints here. We need the returned future to be `Send` otherwise we cannot spawn it onto the executor.

denis2glez replied:
That was my impression too, but I didn't want to add new restrictions to the types from the beginning.
Following this approach we also need some associated types to be `Send`.

thomaseizinger commented on core/src/transport/upgrade.rs:
I am tempted to try and ship this as a non-breaking change. Can you add an `authenticate2` function instead of modifying the current one?

denis2glez replied:
That makes sense, let's add `authenticate2` as an alternative code path.

thomaseizinger reviewed (commented):
Great start!

We'll need to implement the new trait for `libp2p-tls` and` libp2p-noise`.

denis2glez committed:
fix: introduce `Builder::authenticate2` as an alternative code path

Introduce `Builder::authenticate2` as an alternative code path given
the supplied upgrade implements `SecurityUpgrade` instead of
`InboundConnectionUpgrade`/`OutboundConnectionUpgrade`.

Inline `secure_inbound` and `secure_outound` functions into `secure`.

Fix `PeerId` in the output of `SecurityUpgrade::upgrade_security`.

Remove the unnecessary call to `panic`.

Co-authored-by: Thomas Eizinger

denis2glez committed:
docs: improve comments related to `Builder::authenticate2`

thomaseizinger commented on core/src/transport/upgrade.rs:
```diff
@@ -113,6 +113,44 @@ where
+    /// This function is an alternative code path to [`Builder::authenticate`]
+    /// given the supplied upgrade implements `SecurityUpgrade` instead of
+    /// `InboundConnectionUpgrade`/`OutboundConnectionUpgrade`.
+    pub fn authenticate2<C, D, U, E>(
```
Let's use this new function througout the codebase. If we deprecate the old one, we can easily find all usage sites. 

I don't think we want to actually merge a deprecation of `authenticate` though. It might be better to mark this one as `#[doc(hidden)]` and only change the trait-bounds of `authenticate` in the next breaking release.

denis2glez replied:
Sure, no problem.

thomaseizinger commented on core/src/upgrade.rs:
```diff
@@ -152,3 +154,22 @@ pub trait OutboundConnectionUpgrade<T>: UpgradeInfo {
 }
+
+/// Possible security upgrade on a connection
+pub trait SecurityUpgrade<T>: UpgradeInfo {
```
Can you implement this in `libp2p-tls` and `libp2p-noise` please?

thomaseizinger replied:
I think we may also have to split this one into two traits, otherwise the concrete security upgrade doesn't know, which role it is in (dialer vs listener).

denis2glez replied:
Yes, after removing the panic, now we cannot differentiate based on the variant of the `PeerId` parameter.

denis2glez replied:
I implemented the new traits in `libp2p-tls` and `libp2p-noise`.

denis2glez committed:
fix: add `Send` constraint to `Builder::authenticate2` parameters

Use an owned dynamically typed `Future` that is `Send`
(i.e. `BoxFuture`).

denis2glez committed:
refactor: split `SecurityUpgrade` into two alternative traits

Introduce `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` instead.

denis2glez committed:
docs: mark items related to `Builder::authenticate2` as `#[doc(hidden)]`

Plus fix doc comments.

denis2glez committed:
feat: impl `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` for TLS

Implement the `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` traits
for the TLS transport verifying the expected peer ID on outgoing upgrades.

denis2glez committed:
feat: impl `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` for Noise

Implement the `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` traits
for the Noise transport verifying the expected peer ID on outgoing upgrades.

thomaseizinger commented on core/src/upgrade.rs:
```diff
@@ -152,3 +154,39 @@ pub trait OutboundConnectionUpgrade<T>: UpgradeInfo {
+    /// The `info` is the identifier of the protocol, as produced by `protocol_info`. Security
+    /// transports use the optional `peer_id` parameter on outgoing upgrades to validate the
+    /// expected `PeerId`.
+    fn secure_inbound(self, socket: T, info: Self::Info, peer_id: Option<PeerId>) -> Self::Future;
```
We never know the `PeerId` of an inbound connection before we make the security upgrade so we can remove the parameter here.

denis2glez replied:
Of course, it makes sense now.

thomaseizinger commented on transports/noise/src/lib.rs:
```diff
@@ -219,6 +223,59 @@ where
+
+            let (pk, io) = state.finish()?;
+
+            let expected = pk.to_peer_id();
```
This is a bit odd, I would have thought that the provided `PeerId` is to be named `expected` and the one we negotiated is the one we "found".

denis2glez replied:
I definitely looked at it from another perspective :-)

thomaseizinger replied:
When we dial an address, say `/ip4/10.0.0.1/tcp/9999/p2p/XXX` then we _expect_ the peer at the IP+port endpoint to authenticate as `XXX`.

`XXX` is what we extract from the address and pass into the `secure_outbound` function. The public key (thus `PeerId`) we receive as part of the noise handshake (`pk` above) is what the peer _actually_ authenticated with. In other words, they have control over the corresponding private key.

We need to fail the handshake if our expectation is not met. This could either be because somebody is performing a MITM attack or the address is simply wrong (e.g. bad user input). In any case, the `PeerId` coming from the _address_ is the expected `PeerId` and the one from the handshake is the actual `PeerId`.

denis2glez replied:
Thank you for your comments and excuse me for the delay in the response. Definitely the previous wording for the actual and *expected* `PeerId` was a bit confusing, so I preferred to change it to just `peer_id` for the actual one and `remote_peer_id` for the one coming from the address. I noticed that in other parts of the project the latter is already used for a similar purpose.

thomaseizinger replied:
> I noticed that in other parts of the project the latter is already used for a similar purpose.

Whilst we definitely use `remote_peer_id` in other parts of the project, it is not the same "purpose" as it is here. In fact, as part of a security handshake, the _expected_ peer ID and the one the peer authenticated with as part of the handshake are **both** a "remote peer id". So it does not make any sense to use "remote" as a qualifier here.

Please change the variable names to what I suggested:

- The one provided via the parameter is the _expected_ `PeerId` (if you wanted to be super precise, you could call it `maybe_expected_peer_id` to also cover the fact that it is an `Option`)
- The one that is output as part of the auth handshake is the _actual_ `PeerId`

I am open to alternative names but they must convey the same semantics as "expected" and "actual". We are essentially performing a test here and "expected" + "actual" are very common names to use in tests for the exact same purpose (i.e. a failure if they don't match) so I think they are fine names to use for this case.

thomaseizinger commented on transports/tls/src/upgrade.rs:
```diff
@@ -118,6 +123,63 @@ where
+                    Err(UpgradeError::PeerIdMismatch { expected, found })
+                }
+                _ => Ok((expected, stream.into())),
+            }
```
We don't need to do this. Instead, we can make the client and server config ad-hoc in here and pass the expected `PeerId` into `crate::make_client_config`.

denis2glez replied:
Thanks, I'll get into this tomorrow.

thomaseizinger reviewed (commented):
That looks great!

I've left some more comments. Can you also please use `authenticate2` within the `SwarmBuilder` in the `libp2p` crate?

denis2glez committed:
chore: minor doc comments fixes

denis2glez committed:
refactor: remove the `PeerId` parameter in `InboundSecurityUpgrade`

The optional `PeerId` parameter in a security upgrade is never known
at this point for an inbound connection.

thomaseizinger commented on core/src/transport/upgrade.rs:
```diff
@@ -113,6 +114,53 @@ where
+    /// ## Alternative state
+    ///
+    /// This function is an alternative code path to [`Builder::authenticate`]
+    /// given the supplied upgrade implements `InboundConnectionUpgrade`/`OutboundSecurityUpgrade`
```
```suggestion
    /// given the supplied upgrade implements `InboundSecurityUpgrade`/`OutboundSecurityUpgrade`
```

Shouldn't it be this?

Also, can we make these rustdoc links please using `[]` to ensure they get checked at compile-time!

denis2glez replied:
Yes, I missed that :smile:

thomaseizinger commented on core/src/transport/upgrade.rs:
```diff
@@ -113,6 +114,53 @@ where
         ))
     }
+
+    /// Upgrades the transport to perform authentication of the remote
```
```suggestion
    /// Upgrades the transport to perform authentication of the remote.
```

thomaseizinger commented on core/src/upgrade.rs:
```diff
@@ -152,3 +154,37 @@ pub trait OutboundConnectionUpgrade<T>: UpgradeInfo {
+    /// The `info` is the identifier of the protocol, as produced by `protocol_info`. Security
+    /// transports use the optional `peer_id` parameter on outgoing upgrades to validate the
+    /// expected `PeerId`.
+    fn secure_outbound(self, socket: T, info: Self::Info, peer_id: Option<PeerId>) -> Self::Future;
```
```suggestion
    fn secure_outbound(self, socket: T, info: Self::Info, expected_peer_id: Option<PeerId>) -> Self::Future;
```

thomaseizinger commented on core/src/upgrade/secure.rs:
```diff
@@ -0,0 +1,107 @@
+//! After a successful protocol negotiation as part of the upgrade process, the `InboundConnectionUpgrade::secure_inbound`
```
```suggestion
//! After a successful protocol negotiation as part of the upgrade process, the [`InboundSecurityUpgrade::secure_inbound`]
```

thomaseizinger commented on transports/noise/src/lib.rs:
```diff
@@ -58,13 +58,17 @@
 mod protocol;
 
+use futures::future::BoxFuture;
 pub use io::Output;
```
Nit: Let's keep this with the other, non-pub `use` please.

denis2glez replied:
:+1:

thomaseizinger commented on transports/noise/src/lib.rs:
```diff
@@ -219,6 +223,59 @@ where
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_outbound(self, socket: T, _: Self::Info, peer_id: Option<PeerId>) -> Self::Future {
```
Eventually, we will remove the other implementation. Thus, I think to reduce duplication, we can implement `InboundConnectionUpgrade` by delegating to `InboundSecurityUpgrade` and equivalent for outbound.

denis2glez replied:
Already done.

thomaseizinger commented on core/src/upgrade.rs:
```diff
@@ -63,19 +63,21 @@ mod either;
-pub use crate::Negotiated;
+use libp2p_identity::PeerId;
 pub use multistream_select::{NegotiatedComplete, NegotiationError, ProtocolError, Version};
+pub(crate) use secure::{secure, EitherSecurityFuture};
```
Can we please keep the grouping of `pub use`, `pub(crate)` and `use`.

denis2glez replied:
Of course, no problem.

thomaseizinger reviewed (commented):
Good progress! I've left a few more comments :)

denis2glez committed:
docs: fix several issues in the comments

denis2glez committed:
refactor: delegate `upgrade_inbound` and `upgrade_outbound` methods

Delegate the implementations of `InboundConnectionUpgrade`/`OutboundConnectionUpgrade`
traits to `InboundSecurityUpgrade`/`OutboundSecurityUpgrade` for Noise and TLS.

denis2glez committed:
fix: references to actual and remote `PeerId` respectively

The `PeerId` coming from the address is the remote `PeerId` and the one from the handshake is the actual `PeerId`.

denis2glez committed:
chore: keep grouping of module and use declarations

denis2glez committed:
fix: make client and server config ad-hoc in `secure_outbound` method

Introduce the `with_remote_peer_id` constructor in `Config`.

thomaseizinger commented on transports/tls/src/upgrade.rs:
```diff
@@ -57,6 +65,17 @@ impl Config {
+            server: crate::make_server_config(&identity)?,
+            client: crate::make_client_config(&identity, remote_peer_id)?,
+        })
+    }
```
Instead of having this, I'd prefer if we instead store the `identity::Keypair` within `Config` and call the `make_server_config` and `make_client_config` as part of the upgrade.

denis2glez replied:
Done :+1:

denis2glez committed:
fix: use the wording of `actual` and `expected` PeerId

The `PeerId` coming from the address is the expected `PeerId`
and the one from the handshake is the actual `PeerId`.

denis2glez committed:
refactor: store the `identity::Keypair` within `Config`

Explicitly call `make_server_config` and `make_client_config` as part of the upgrade.
